[#5360] Add iphymv exit code tests (master)

alanking opened the pull request:
Still need to run the mungefs tests, but here is the start. Companion PR is irods/irods_client_icommands#201

trel commented:
very nice.    i see all seven error codes.

alanking commented:
I added more assertions to the `iphymv` tests as well as some checksum mismatch tests for `icp`, `iput`, and `irepl` to test behavior in those API endpoints as well. This uncovered an issue in `icp` where error codes in finalize were not being propagated back to the user (although the replica was correctly staled). I need to do another CI run but my Jenkins is down, so might take a while to get there.

trel commented on plugins/database/src/db_plugin.cpp:
```diff
@@ -2992,6 +2994,19 @@ irods::error db_reg_replica_op(
+    // The caller has requested that the replica be registered as intermediate. This
+    // means that the replica will be written or changed at a future time. Otherwise,
+    // the replica will take the replica status of the source replica.
+    char* intermediate_replica_status = "2";
```
this isn't defined somewhere else already as '2'?

alanking replied:
What I normally do is a `std::to_string` on the `INTERMEDIATE_REPLICA` enumerated value, but that won't play well here as `cVal` is an array of pointer to non-const `char`s. I could cast away the const or perform a `std::snprintf` to a pre-existing char array. Casting away the const seems pretty safe given it's coming from a temporary anyway.

trel commented on server/api/src/rsDataObjUnlink.cpp:
```diff
@@ -547,125 +564,95 @@ int dataObjUnlinkS(rsComm_t* rsComm,
+                "[error_code=[{}], path=[{}], hierarchy=[{}]",
+                __FUNCTION__, __LINE__, ec, info.logical_path(), info.hierarchy()));
+
+            // TODO: why do we allow NOENT and EACCES to unregister?
```
b/c the physical file already has a problem?   NOENT means it's not even there... and EACCES means the service account can't read it anyway...

alanking replied:
I think I will revisit this on the pass to prevent unlinks on locked data objects.

trel commented on server/api/src/rsDataObjUnlink.cpp:
```diff
@@ -547,125 +564,95 @@ int dataObjUnlinkS(rsComm_t* rsComm,
-                status = 0;
-            }
-        }
+        // TODO: Should the replica be marked stale if the unregister itself fails?
```
if the data on disk and the entry in the catalog are known to be good... we should leave it good.   for *anything* other case, we can't vouch for it, so yes, stale it.

alanking replied:
I think I will revisit this on the pass to prevent unlinks on locked data objects.

alanking commented:
Force push was from an accidental rebase, but no conflicts were found. Addressed comment in db_plugin. Changes before latest commit passed core tests in CI, awaiting unit tests and timing tests.

alanking commented:
Everything passed except for a couple of unit test assertions which I found are likely due to timing issues in updating the catalog on agent teardown. I'm going to address those in the test commit.

alanking commented:
All tests passed in CI. Squashed down commits to RST work, tests, and phymv updates.

trel commented:
in your 5360 commit message ... "propogating" -> "propagating"

everything else looks great.

alanking commented:
Fixed commit message and #'d

trel commented:
awaiting green
